### PR TITLE
BUG: Fix NaT handling in the PyArray_CompareFunc for datetime and timedelta

### LINKED ISSUE #19574: PyArray_CompareFunc of dtype datetime64 returns incorrect value for 'NaT' objects.
Sort order for `NaT` values got [changed in numpy 1.18.0](https://numpy.org/doc/stable/release/1.18.0-notes.html#nat-now-sorts-to-the-end-of-arrays).

`PyArray_CompareFunc` of dtype `datetime64` returns incorrect as per the updated sort order.

This is present with numpy 1.20.2 and 1.21 too.

This is reproducible via the following extension module.

compare_bug.c
```c
#include <Python.h>
#include "numpy/arrayobject.h"
#include "numpy/ndarrayobject.h"
#include "numpy/ndarraytypes.h"
#include "numpy/arrayscalars.h"
#include "numpy/ufuncobject.h"


static PyObject*
reproduce_bug(PyObject* self, PyObject* args) {
    PyObject *arr = NULL;
    if (!PyArg_ParseTuple(args, "|O", &arr)) {
        return Py_None;
    }
    PyArrayObject *numpy_arr = (PyArrayObject*) arr;
    // We just compare 0th and 1st element and print out
    // the returned value.
    int x = numpy_arr->descr->f->compare(
        PyArray_GETPTR1(numpy_arr, 0),
        PyArray_GETPTR1(numpy_arr, 1),
        numpy_arr
    );
    printf("DEBUG %d\n", x);
    return Py_None;
}

static PyMethodDef bug_functions[] = {
    {"bug", (PyCFunction)reproduce_bug, METH_VARARGS, NULL},
    { NULL, NULL, 0, NULL }
};

static struct PyModuleDef numpybugmodule = {
    PyModuleDef_HEAD_INIT,
    "numpy_bug",
    NULL,
    -1,   
    bug_functions
};

PyMODINIT_FUNC
PyInit_numpy_bug(void)
{
    return PyModule_Create(&numpybugmodule);
}
```

setup.py
```python
from distutils.core import setup, Extension, DEBUG
import numpy

extra_compile_args = ["-DNDEBUG", "-O3"]
sfc_module = Extension('numpy_bug', sources = ['compare_bug..c'],
                        include_dirs=[numpy.get_include()],
                       extra_compile_args=extra_compile_args)

setup(name = 'mtnumpy', version = '1.0',
    description = 'Python Package with superfastcode C++ extension',
    ext_modules = [sfc_module]
    )
```

Running this after installing above:
```python
In [1]: import numpy, numpy_bug

In [2]: numpy_bug.bug(numpy.array(['NaT', '2020-01-01'], dtype='datetime64[s]')) # should print 1 as per new sort order.
DEBUG -1
```

### PR DESCRIPTION
**Pull request by asmeurer**
In #12658 and #15068 the sort ordering for datetime and timedelta was changed
so that NaT sorts to the end, but the internal compare array function was not
updated.

Fixes #19574.

If a test should be added for this, please let me know where.

### CODE REVIEW

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 1]
```diff
@@ -2920,6 +2918,43 @@ C@TYPE@_compare(@type@ *pa, @type@ *pb)
 
 /**end repeat**/
 
+/**begin repeat
+ * #TYPE = DATETIME, TIMEDELTA#
+ * #type = npy_datetime, npy_timedelta#
+ */
+
+#define LT(a,b) ((a) < (b) || ((b) != (b) && (a) ==(a)))
```
Comment (charris) [inline numpy/core/src/multiarray/arraytypes.c.src:2926]: I don't think you need this, the `((b) != (b) && (a) ==(a)))` part is for dealing with nans, which isn't a problem with integer types like `npy_datetime` and `npy_timedelta`.

File: numpy/core/src/multiarray/arraytypes.c.src [hunk 2]
```diff
@@ -2920,6 +2918,43 @@ C@TYPE@_compare(@type@ *pa, @type@ *pb)
...
+static int
+@TYPE@_compare(@type@ *pa, @type@ *pb)
+{
+    const @type@ a = *pa;
+    const @type@ b = *pb;
+    int ret;
+
+    printf("Calling @TYPE@_compare\n");
+    if (a == NPY_DATETIME_NAT) {
+        ret = 1;
```
Comment (charris) [inline numpy/core/src/multiarray/arraytypes.c.src:2937]: What if both `a` and `b` are NaT? I note that sort order is not necessarily the same as comparison, which is one of the reasons only `LT`  is used for sorting.

File: numpy/core/src/multiarray/arraytypes.c.src (hunk 1 above)
Comment (asmeurer) [reply numpy/core/src/multiarray/arraytypes.c.src:2926]: Ah you are right. I did notice that I copied the float template instead of the integer template.

**Comment by charris** [comment]
Thanks @asmeurer . This could probably be tested using an object array containing time* types, something like
```
In [20]: a = array([timedelta64('NaT'), timedelta64(0)], dtype=object)          

In [21]: a.sort() 
```
But that is hard to check without merging this.

**Pull request merged**